### PR TITLE
Fix dtype issue in stirling approximation

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes the following repro:
```python
python -c "import jax; jax.random.binomial(jax.random.key(0), jax.numpy.float16(1), jax.numpy.float16(0.5))"
```
Related to https://github.com/jax-ml/jax/pull/25688#discussion_r1938010569

### CODE REVIEW

**Comment by carlosgmartin** [comment]
Just out of curiosity, what's the advantage to doing this over using [`jax.numpy.clip`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.clip.html) in this context?

**Comment by jakevdp** [comment]
`clip` in this context would fail when run with `jax_numpy_dtype_promotion='strict'`. If we want users to be able to set this for their own code, we need to satisfy it for library code.

**Review by sharadmv — APPROVED** [review]
